modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java

```diff
@@ -241,19 +242,29 @@ private void onReplicaMessageReceived(NetworkMessage message, ClusterNode sender
         // If the request actually came from the network, we are already in the correct thread that has permissions to do storage reads
         // and writes.
         // But if this is a local call (in the same Ignite instance), we might still be in a thread that does not have those permissions.
-        if (currentThreadCannotDoStorageReadsAndWrites()) {
+        if (shouldSwitchToRequestsExecutor()) {
             requestsExecutor.execute(() -> handleReplicaRequest(request, sender, correlationId));
         } else {
             handleReplicaRequest(request, sender, correlationId);
         }
     }
 
-    private boolean currentThreadCannotDoStorageReadsAndWrites() {
+    private static boolean shouldSwitchToRequestsExecutor() {
         if (Thread.currentThread() instanceof ThreadAttributes) {
             ThreadAttributes thread = (ThreadAttributes) Thread.currentThread();
             return !thread.allows(STORAGE_READ) || !thread.allows(STORAGE_WRITE);
         } else {
-            // We don't know for sure.
+            if (PublicApiThreading.executingSyncPublicApi()) {
+                // It's a user thread, it executes a sync public API call, so it can do anything, no switch is needed.
+                return false;
+            }
+            if (PublicApiThreading.executingAsyncPublicApi()) {
+                // It's a user thread, it executes an async public API call, so it cannot do anything, a switch is needed.
+                return true;
+            }
+
+            // It's something else: either a JRE thread or an Ignite thread not marked with ThreadAttributes. In any case,
```

Comment:
Should we throw an assertion here immediately?

Reply:
Why? If thread assertions are enabled and the thread touches the storage, we'll get the same assertion, just with a deeper stack trace.